internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -632,6 +632,53 @@ func (r *BMACReconciler) findInstallationDiskID(devices []aiv1beta1.HostDisk, hi
 	return ""
 }
 
+// Finds the agents related to this ClusterDeployment
+//
+// The ClusterDeployment <-> agent relation is one-to-many.
+// This function returns all Agents whose ClusterDeploymentName name
+// matches this ClusterDeployment's name.
+func (r *BMACReconciler) findAgentsByClusterDeployment(ctx context.Context, clusterDeployment *hivev1.ClusterDeployment) []*aiv1beta1.Agent {
+	agentList := aiv1beta1.AgentList{}
+	err := r.Client.List(ctx, &agentList)
+	if err != nil {
+		return nil
+	}
+
+	// There may be more than one Agent that maps to the same BMH
+	// if the cluster deployment had previous failed installations.
+	// Only the newest agent for each BMH is returned.
+	bmhToAgentMap := map[string]*aiv1beta1.Agent{}
+	for i, agent := range agentList.Items {
+		if agent.Spec.ClusterDeploymentName == nil {
+			continue
+		}
+		if agent.Spec.ClusterDeploymentName.Name != clusterDeployment.Name {
+			continue
+		}
+		if bmhName, ok := agent.ObjectMeta.Labels[AGENT_BMH_LABEL]; ok {
+			if existingAgent, ok := bmhToAgentMap[bmhName]; ok {
+				// if there are two Agent with the same bmhName, return the newest
+				if agent.ObjectMeta.CreationTimestamp.After(existingAgent.ObjectMeta.CreationTimestamp.Time) {
+					bmhToAgentMap[bmhName] = &agentList.Items[i]
+				}
+			} else {
+				bmhToAgentMap[bmhName] = &agentList.Items[i]
+			}
+		}
+	}
+
+	agents := []*aiv1beta1.Agent{}
+	for _, agent := range bmhToAgentMap {
+		agents = append(agents, agent)
+	}
+
+	if len(agents) == 0 {
+		return nil
+	}
```

Comment:
Is this check needed or is it possible to just return `agents` which would be an empty slice? dito in L644